src/main.cpp

```diff
@@ -3341,7 +3334,18 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
 
         // Connect new blocks.
         BOOST_REVERSE_FOREACH(CBlockIndex *pindexConnect, vpindexToConnect) {
-            if (!ConnectTip(state, chainparams, pindexConnect, pindexConnect == pindexMostWork ? pblock : NULL)) {
+            const CBlock* pconnectBlock;
+            CBlock block;
+            if (pblock && pindexConnect == pindexMostWork) {
```

Comment:
@str4d Adding a check that `pblock != NULL` here is necessary, I think, but perhaps not sufficient; it's probably important to check that pblock is also actually related to pindexMostWork - we have a comment at the start of the procedure that says that should be respected, but then nothing that actually enforces that invariant?

Reply:
Enforcing this invariant would be nice, but that's a much larger refactor. This is a good separate step.